Source/Model/Classes/NMR_Model.cpp

```diff
@@ -391,19 +391,30 @@ namespace NMR {
 		}
 	}
 
-
-	// Retrieve a unique Resource ID
-	ModelResourceID CModel::generateResourceID()
-	{
-		// TODO: is this truly safe?
-		auto iIterator = m_ResourceMap.rbegin();
-		if (iIterator != m_ResourceMap.rend())
-			return iIterator->first + 1;
-		else
-			return 1;
-	}
-
-	void CModel::updateUniqueResourceID(UniqueResourceID nOldID, UniqueResourceID nNewID)
+    ModelResourceID CModel::generateResourceID()
+    {
+        // Build a map with existing resource IDs
+        std::map<ModelResourceID, bool> existingResourceIDs;
```

Comment:
this could be a std::unordered_set. since we dont use the bool value.

Reply:
I fear this is a major bottleneck as it will build up the map every time "generateresourceID" is called.
It would be much easier to store a m_MaxResourceID in the model that is updated whenever we add a resource.